### PR TITLE
Fix E2E test failure

### PR DESCRIPTION
**Pull request by jianrongzhang89**
OSD E2E prow job failed:
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/osde2e-stage-aws-addon-dbaas-operator/1541753408988385280
Error message:
admission webhook "vdbaasinventory.kb.io" denied the request: DBaaSProvider.dbaas.redhat.com "XXXXXXXXXXXXXXXXXXXXXXXXXXX" not found
  occurred[0m
  [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/go/src/github.com/RHEcosystemAppEng/dbaas-e2e-test-harness/pkg/tests/dbaas_operator_tests.go:135[0m]

This PR is to check if the provider operator has finished its registration with the DBaaS operator by creating its dbaasprovider CR.

Tested done on the lab https://api.rhoda-49-lab.5xr9.p1.openshiftapps.com:6443

### CODE REVIEW

File: pkg/tests/dbaas_operator_tests.go [hunk 1]
```diff
@@ -3,8 +3,13 @@ package tests
...
 	"context"
 	"errors"
-	"flag"
 	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
```
Comment (bmozaffa) [inline pkg/tests/dbaas_operator_tests.go]: Curious why these are getting rearranged

File: pkg/tests/dbaas_operator_tests.go [hunk 2]
```diff
@@ -54,6 +54,18 @@ var _ = Describe("Rhoda e2e Test", func() {
 			//loop through providers
 			for i := range providers {
 				provider := providers[i]
+				providerType := string(provider.SecretData["providerType"])
+				BeforeEach(func() {
```
Comment (bmozaffa) [inline pkg/tests/dbaas_operator_tests.go:57]: Given that each loop has a single (It) subject node, I assume the BeforeEach will run just once in each iteration? What is the impact of using BeforeEach in this case, versus doing the check in each subject node?

File: pkg/tests/dbaas_operator_tests.go (hunk 1 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go]: This was done by IDE (Visual Studio Code) using gofmt automatically so that the entries are organized in a standard way (core modules, 3rd party modules, and then modules of the current project).

File: pkg/tests/dbaas_operator_tests.go (hunk 2 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go:57]: Here BeforeEach is executed for the "It" once for each provider. If "It" were replaced with "DescribeTable" which has multiple test cases, then BeforeEach would run once for each test case, for instance:
https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/api/v1alpha1/dbaasconnection_webhook_test.go#L49

About: "What is the impact of using BeforeEach in this case, versus doing the check in each subject node?", BeforeEach() is designed to be used right before "It". We could not put it inside "It", unless it contains child test cases in which it would run once for each child test case.

File: pkg/tests/dbaas_operator_tests.go (hunk 1 above)
Comment (bmozaffa) [reply pkg/tests/dbaas_operator_tests.go]: The problem with that is that we haven't standardized on VS Code, and if anything, I would personally much prefer IntelliJ. So we risk having ping pong and a lot of unnecessary chatter on the diff. So let's either revert, or make it part of the build script.

File: pkg/tests/dbaas_operator_tests.go (hunk 2 above)
Comment (bmozaffa) [reply pkg/tests/dbaas_operator_tests.go:57]: To recap and make sure I understand:

In the current setup, there is no difference between doing BeforeEach and having it as a step inside the subject node.

If the code is refactor to use multiple specs for each provider, BeforeEach would run multiple times for each provider, which is NOT what we want.

So not a good fit?

File: pkg/tests/dbaas_operator_tests.go (hunk 1 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go]: Good point. I updated the Makefile so that the build will check the format first. It uses goimports that is commonly used in the community.

File: pkg/tests/dbaas_operator_tests.go (hunk 2 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go:57]: In that case, when the code is refactored so that each provider has multiple specs, we should put these specs inside a single "Context" and then the BeforEach would run only once for each provider.

File: Makefile [hunk 3]
```diff
@@ -3,7 +3,13 @@ IMG ?= quay.io/ecosystem-appeng/dbaas-e2e-test-harness:latest
 DIR := $(dir $(realpath $(firstword $(MAKEFILE_LIST))))
 OUT_FILE := "$(DIR)dbaas-e2e-test-harness"
 
-build:
+install-tools:
+	go get golang.org/x/tools/cmd/goimports
```
Comment (bmozaffa) [inline Makefile]: > go get: installing executables with 'go get' in module mode is deprecated.

File: Makefile [hunk 4]
```diff
@@ -3,7 +3,13 @@ IMG ?= quay.io/ecosystem-appeng/dbaas-e2e-test-harness:latest
 DIR := $(dir $(realpath $(firstword $(MAKEFILE_LIST))))
 OUT_FILE := "$(DIR)dbaas-e2e-test-harness"
 
-build:
+install-tools:
```
Comment (bmozaffa) [inline Makefile]: This makes install-tools the default build target... which we don't want

File: pkg/tests/dbaas_operator_tests.go (hunk 1 above)
Comment (bmozaffa) [reply pkg/tests/dbaas_operator_tests.go]: This does not format the code, but instead fails the build if not formatted according to what it expects, and without providing a good / clear message. That's an issue for IntelliJ users among us!

File: pkg/tests/dbaas_operator_tests.go (hunk 2 above)
Comment (bmozaffa) [reply pkg/tests/dbaas_operator_tests.go:57]: That doesn't work, and afaiu, Context() is irrelevant. It() blocks define test specs and BeforeEach means run once for each of them. So again, I'm looking for any good reason to use BeforeEach() here. If it was "Before()", which unfortunately does not exist, then it would make sense. But I don't think this does.

File: pkg/tests/dbaas_operator_tests.go [hunk 5]
```diff
@@ -150,10 +162,9 @@ var _ = Describe("Rhoda e2e Test", func() {
 								}
 							}
 							return false
-						} else {
-							fmt.Println("inventory.Status.Conditions Len is 0")
-							return false
 						}
+						fmt.Println("inventory.Status.Conditions Len is 0")
+						return false
```
Comment (bmozaffa) [inline pkg/tests/dbaas_operator_tests.go]: Not sure why this change was made. I find the previous format much more readable. You know that the log statement is printed in the else, if Conditions Len is NOT > 0... explicit If/Else makes it easier to understand what goes on, versus an implicit else because the if ends with a return.

File: pkg/tests/dbaas_operator_tests.go [hunk 6]
```diff
@@ -194,10 +205,9 @@ var _ = Describe("Rhoda e2e Test", func() {
 									}
 								}
 								return false
-							} else {
-								fmt.Println("dbaaSConnection.Status.Conditions Len is 0")
-								return false
```
Comment (bmozaffa) [inline pkg/tests/dbaas_operator_tests.go]: Ditto

**Review by bmozaffa — CHANGES_REQUESTED** [review]
This is getting a bit too complicated... please do a PR with only the provider wait change, and do a separate PR for any code/build improvements you have in mind.

File: pkg/tests/dbaas_operator_tests.go (hunk 5 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go]: This is actually a common practice for go and is inforced by golint: when the if statement ends with a "return", the "else" statement is redundant so for simplicity it is removed here. 

Anyway, I will update this PR to only include the changes for adding the BeforeEach(), and all format-related changes will be done by a separate PR.

File: pkg/tests/dbaas_operator_tests.go (hunk 2 above)
Comment (jianrongzhang89) [reply pkg/tests/dbaas_operator_tests.go:57]: It will be something like below - note that multiple "It"s are inside a new Context().
				BeforeEach(func() {
					By(fmt.Sprintf("Checking if provider operator is ready: %s", provider.ProviderName))
					Eventually(func() bool {
						if err := client.Get(context.Background(), k8sClient.ObjectKey{
							Name: providerType,
						}, &dbaasv1alpha1.DBaaSProvider{}); err != nil {
							return false
						}
						return true
					}, time.Second*120, time.Second).Should(BeTrue())
				})
				Context("Test provider cases", func() {
					It("Should pass when secret and provider created and connection status is checked for "+provider.ProviderName, func() { ....})
					It("Should test case B pass, func() { ....})
					It("Should test case C pass, func() { ....})
                                  })

**Review by bmozaffa — APPROVED** [review]

**Comment by bmozaffa** [comment]
Thanks, @jianrongzhang89 , this has been merged, built and pushed. The next [prow job](https://prow.ci.openshift.org/?job=osde2e-stage-aws-addon-dbaas-operator) at 8am your time should pick it up, so looking forward to the results.

Please do move forward with a second PR for your other improvements next.
